pkg/querier/queryrange/stats.go

```diff
@@ -46,31 +87,57 @@ func statsMiddleware(recorder metricRecorder) queryrange.Middleware {
 
 			// collect stats and status
 			var statistics *stats.Result
-			var status string
 			if resp != nil {
 				switch r := resp.(type) {
 				case *LokiResponse:
 					statistics = &r.Statistics
-					status = r.Status
 				case *LokiPromResponse:
 					statistics = &r.Statistics
-					if r.Response != nil {
-						status = r.Response.Status
-					}
 				default:
 					level.Warn(util.Logger).Log("msg", fmt.Sprintf("cannot compute stats, unexpected type: %T", resp))
 				}
 			}
-			if err != nil {
-				status = loghttp.QueryStatusFail
-			}
 			if statistics != nil {
 				// Re-calculate the summary then log and record metrics for the current query
 				statistics.ComputeSummary(time.Since(start))
 				statistics.Log(logger)
-				recorder.Record(status, req.GetQuery(), logql.GetRangeType(paramsFromRequest(req)), *statistics)
+			}
+			ctxValue := ctx.Value(ctxKey)
+			if data, ok := ctxValue.(*queryData); ok {
+				data.recorded = true
+				data.query = req.GetQuery()
+				data.statistics = statistics
+				data.rangeType = logql.GetRangeType(paramsFromRequest(req))
 			}
 			return resp, err
 		})
 	})
 }
+
+// interceptor implements WriteHeader to intercept status codes. WriteHeader
+// may not be called on success, so initialize statusCode with the status you
+// want to report on success, i.e. http.StatusOK.
+//
+// interceptor also implements net.Hijacker, to let the downstream Handler
+// hijack the connection. This is needed, for example, for working with websockets.
+type interceptor struct {
+	http.ResponseWriter
+	statusCode int
+	recorded   bool
+}
+
+func (i *interceptor) WriteHeader(code int) {
+	if !i.recorded {
```

Comment:
Are multiple things calling WriteHeader()?  Is always taking the first one correct?   Possible race condition here if multiple goroutines are handling this http call?

Reply:
This is the same code from weaveworks server. Potentially this could happens you're right but we never call WriteHeader from multiple goroutine, and go handle request in a single goroutine.